### ISSUE TITLE
mkvmerge high memory usage when merging large mkv files

### ISSUE THREAD

**Issue by felixbuenemann** [mbunkus/mkvtoolnix#1893]
I'm seeing extremely high memory usage in mkvmerge 9.8.0 linux x64 when merging two parts of a large UHD mkv.

I had a 60GB mkv that I split up into two parts using `mkvmerge --split 50G`.

The file had the following tracks:

```
0:HEVC/h.265
1:AC-3
2:AC-3
3:TrueHD
4:HDMV PGS
5:HDMV PGS
6:HDMV PGS
```

When merging the 50GB and the 10GB part back together using `mkvmerge --append-mode track -o merged.mkv 50g.mkv + 10g.mkv` mkvmerge progressed quickly to 50% with minimal memory usage (copying the first 50GB), then quickly added the subtitle tracks (I believe these are always appended in "file" append-mode, if that matters) and then it just sat there slowly eating up memory.

The system had only 2 GB RAM, so I ended up adding 12 GB of swap to keep mkvmerge from crashing. 

I monitored the memory usage and after the swap usage and virtual memory usage of mkvmerge had reached 10 GB (with 1.8 GB RSS), mkvmerge printed that it had added track 0, 1, 2, 3 and finished shortly after that, the whole process taking 70 minutes.

I normally run this VM with only 512 MB RAM and 2 GB swap and in that setup it just about managed to merge a file split into a 50GB and 1GB part with using lots of swap.

So it appears that mkvmerge requires as much memory as the merged part.

I could retry the experiment with `--append-mode file` if that is deemed useful.

**Comment by felixbuenemann** [comment]
Btw. is "track" append-mode even the right tool to use for mkv files split by mkvmerge or would the default "file" mode be more appropriate in this case? (I read the docs on that subject, but they didn't make me any wiser.)

**Comment by mbunkus** [comment]
In general mkvmerge shouldn't consume arbitrary amounts of memory. There are certain known cases where this happens, e.g. if mkvmerge is unable to detect key frames in h.264/h.265 tracks.

Please run the following and upload the compressed `output.txt` to [my FTP server](https://github.com/mbunkus/mkvtoolnix/wiki/FTP-server):

    mkvinfo --redirect-output output.txt --summary first_file.mkv

> I could retry the experiment with `--append-mode` file if that is deemed useful.

Using a different append mode wouldn't affect this particular problem.

> Btw. is "track" append-mode even the right tool to use for mkv files split by mkvmerge or would the default "file" mode be more appropriate in this case?

Yes, it is.

**Comment by felixbuenemann** [comment]
Unfortunately I can't connect to the SFTP-Server, because it only supports weak and outdated ciphers: `arcfour256,arcfour128,blowfish-ctr,blowfish-cbc`

I tried with both `/usr/bin/sftp` (OpenSSH_7.3p1, LibreSSL 2.4.1), which supports `chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,3des-cbc` and `/usr/local/bin/sftp` (OpenSSH_7.4p1, OpenSSL 1.0.2k), same ciphers except `3des-cbc`.

You might want to consider at least adding `aes128-cbc` to the list, which should be available even on pretty old servers and is still supported by newer clients ([benchmark of different ciphers](https://blog.famzah.net/2015/06/26/openssh-ciphers-performance-benchmark-update-2015/)).

In the meantime I've uploaded the files to dropbox:

* [issue1893_pt1_output.txt.gz](https://www.dropbox.com/s/24ujdw6i6tmvad9/issue1893_pt1_output.txt.gz?dl=1) (93 MB from first part)
* [issue1893_pt2_output.txt.gz](https://www.dropbox.com/s/03jaew90n2vxwlx/issue1893_pt2_output.txt.gz?dl=1) (19 MB from second part)

**Comment by mkver** [comment]
I have made some tests myself and I think I found the culprit: mkvmerge reads and caches until it has found one packet for each track that is about to be muxed. In Felix' file this is problematic because track 7 (a forced subtitle?) only has two packets near the end (at about 21:05). In the test files that I am currently uploading the second file contains an empty forced subtitle track. If one appends these two files with the forced subtitle track, the issue appears; if one muxes without appending the forced subtitle from the second file (even if one includes the forced subtitles from the first file), the issue does not appear. I also split the file differently than in the files that I uploaded, namely so that there appears a forced subtitle packet at the beginning of the second file. In this case, there is no unusual amount of memory usage during appending.

**Comment by mbunkus** [comment]
@felixbuenemann  Thanks, both for the files and the hint about the sftp configuration. Back in the day I had indeed added that restriction to the configuration, though I don't remember why — probably in order to get some obscure client to work.

I've simply removed the restriction, and the server now offers all supported ciphers:

>  debug2: ciphers stoc: aes256-ctr,aes192-ctr,aes128-ctr,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-ctr,blowfish-cbc,cast128-cbc,arcfour256,arcfour128,3des-ctr,3des-cbc

@mkver Thanks for the analysis.

I'll look into this soonish.

**Comment by felixbuenemann** [comment]
@mbunkus Thanks, sftp access is working fine now.

**Comment by felixbuenemann** [comment]
@mkver Your assumptions are correct, track seven is the forced subtitle track. I checked a text version of the subs and it should contain 5 subtitles in the first part and a single subtitle with the track language at the very ending of the second part.

**Comment by mkver** [comment]
@felixbuenemann And what happens if you don't append the second part of the forced subtitles? Do you still get the high memory consumption?

**Comment by felixbuenemann** [comment]
@mkver I tried `mkvmerge --append-mode track -o out.mkv pt1.mkv -s !6 pt2.mkv` (skipping subtitle track 7 for part two) and memory usage was low, around 75MB.

**Comment by felixbuenemann** [comment]
I can confirm this as fixed in mkvmerge v9.9.0.
